packages/babel-plugin-transform-typescript/src/enum.ts

```diff
@@ -131,109 +137,168 @@ const enumSelfReferenceVisitor = {
 function evaluate(
-  expr: t.Node,
+  path: NodePath,
   seen: PreviousEnumMembers,
-): number | string | typeof undefined {
-  return evalConstant(expr);
+): number | string | undefined {
+  return evalConstant(path);
 
-  function evalConstant(expr: t.Node): number | typeof undefined {
+  function evalConstant(path: NodePath): number | string | undefined {
+    const expr = path.node;
     switch (expr.type) {
+      case "MemberExpression": {
+        const obj = expr.object;
+        const prop = expr.property;
+        if (
+          !types.isIdentifier(obj) ||
+          (expr.computed
+            ? !types.isStringLiteral(prop)
+            : !types.isIdentifier(prop))
+        ) {
+          return;
+        }
+        const bindingIdentifier = path.scope.getBindingIdentifier(obj.name);
+        const data = ENUMS.get(bindingIdentifier);
+        if (!data) return;
+        // @ts-expect-error checked above
+        return data.get(prop.computed ? prop.value : prop.name);
+      }
+
       case "StringLiteral":
         return expr.value;
       case "UnaryExpression":
-        return evalUnaryExpression(expr);
+        return evalUnaryExpression(path as NodePath<t.UnaryExpression>);
       case "BinaryExpression":
-        return evalBinaryExpression(expr);
+        return evalBinaryExpression(path as NodePath<t.BinaryExpression>);
       case "NumericLiteral":
         return expr.value;
       case "ParenthesizedExpression":
-        return evalConstant(expr.expression);
-      case "Identifier":
-        return seen.get(expr.name);
-      case "TemplateLiteral":
+        return evalConstant(path.get("expression"));
+      case "Identifier": {
+        let value = seen.get(expr.name);
+        if (value === undefined) {
+          value = evaluateIdentifier(path);
+          if (value !== undefined) {
+            seen.set(expr.name, value);
+          }
+        }
+        return value;
+      }
+      case "TemplateLiteral": {
         if (expr.quasis.length === 1) {
           return expr.quasis[0].value.cooked;
         }
-      /* falls through */
+
+        const paths = (path as NodePath<t.TemplateLiteral>).get("expressions");
+        const quasis = expr.quasis;
+        let str = "";
+
+        for (let i = 0; i < quasis.length; i++) {
+          str += quasis[i].value.cooked;
+
+          if (i + 1 < quasis.length) {
+            const value = evaluateIdentifier(paths[i]);
+            if (value === undefined) return undefined;
+            str += value;
+          }
+        }
+        return str;
+      }
       default:
         return undefined;
     }
   }
 
-  function evalUnaryExpression({
-    argument,
-    operator,
-  }: t.UnaryExpression): number | typeof undefined {
-    const value = evalConstant(argument);
+  function evaluateIdentifier(path: NodePath) {
+    if (path.isIdentifier()) {
+      const binding = path.scope.getBinding(path.node.name);
+      if (binding?.kind === "const") {
```

Comment:
Why do we need to check if the binding is `const`? If there are `constantViolations`, `.evaluate()` already returns a non-confident result.

Reply:
ts will only inline `const` variables, and `.evaluate()` will treat unassigned variables as confident.
Originally I was doing this to avoid potential bugs due to variable reference information anomalies, as our scope information is often unreliable, but now I'm not sure if this is necessary.
This would avoid https://github.com/babel/babel/pull/15379#issuecomment-1408525848 if inlining all unassigned variables, but would be somewhat inconsistent with ts.
I can't make a choice, what do you think?

Reply:
Is the behavior of inlining all the variables different at runtime from TS? If not, we can inline all of them.

Reply:
Yes.
![image](https://user-images.githubusercontent.com/30521560/215506304-790de4cb-f1b2-4349-aec3-2fd3c7bac34b.png)
https://www.typescriptlang.org/play?declaration=false&isolatedModules=true&ts=5.0.0-dev.20230130#code/MYewdgzgLgBAQgQwgUwGoIDYFdkwLwwCMADANwCwAUKJLAAoBOyAZgJYAe+MARAPQAmCKAm4Vq4aDGRgsAWxjpsyCDADeVGJpgAxVg0kFEKRTgA0m3ryLENWgMrIa-c1tdvLRQrc0AVABZ6-G7Brh6EAExUAL5UNJLScjAASiBYUMpq3jB0CAxQKgQABgAkqowsHFG8AA65+YUu7lZ8gsI1dRDcWQCSYABuIKzAGUWl5WzsVaz9g8MQhRbNAkIIvNMDQ8pdlDGUGMiwRmiYOOFcJGJxsAnyJspn6pSuuvqwhkjHSuEuYTZP9o5wM4QsEwl5-r4AgwgiDQVYItEgA

Reply:
I lean towards always inlining them anyway, since the only observable difference is for code where TS would report an error anyway (i.e. when all the enum members must have an initializer but some don't), but let's also see what @JLHwung thinks.

Reply:
I also found another inconsistency, I opened https://github.com/microsoft/TypeScript/issues/52510.

Reply:
I agree with @nicolo-ribaudo. Generally the transform plugin should assume the AST is correct, otherwise there would have been a parsing error. And in this case Babel parser should ensure that we don't have enum members without initializers.